### PR TITLE
Adds support for asynchronous executions via Promises

### PR DESCRIPTION
**Pull request by acelot**
- New `promiseExecute` method in `Execution.php`
- Helpers for API (`graphqlAsync()`, `processAsync`)

### CODE REVIEW

**Comment by Jalle19** [comment]
Would it be possible to use the async version all the time, and just change the `graphql` helper function to wait for the result? That way there would be a lot less code duplication.

**Comment by acelot** [comment]
@Jalle19 I think so too, but did so for backward compatibility

**Comment by Jalle19** [comment]
> @Jalle19 I think so too, but did so for backward compatibility

You can think of the `graphql()` helper function and it's related functions as the public API, not the actual public interfaces etc.

**Comment by acelot** [comment]
It seems to me that synchronous and asynchronous methods should exist separately, since impossible to turn asynchronous promise into synchronous, because the result of promise comes through a callback.

**Comment by Jalle19** [comment]
Isn't it possible to just add `->then()` and return the result to turn it into synchronous?

**Comment by acelot** [comment]
Nope. In any case, will return Promise.

```php
return GraphQL::promiseExecute(...)->then(function (ExecutionResult $result) {
    return $result;
})
```

```php
$result = graphql(...);
var_export($result instanceof Promise); // true
```

**Comment by Jalle19** [comment]
Isn't there some wait method that could be used?

**Comment by acelot** [comment]
All promises works in synchronous way without event loop. But thanks to the ReactPHP event loop PHP works in real asynchronous mode. Unfortunately I don't know how ReactPHP resolves promises.

**Comment by acelot** [comment]
I have refactored the code. Now `execute` method in `Execution` always returns a Promise. I have moved promise resolving for synchronous calls into API method `graphql`. I also added a check, if a Event Loop is used (React PHP), an error is returned.

**Comment by acelot** [comment]
Any chance that this PR will be merged?

**Comment by Jalle19** [comment]
I'll try to find some time to test this out this week, the latest changes look very promising (no pun intended)!

**Comment by Jalle19** [comment]
@acelot I fixed the tests, for some reason Travis CI doesn't run on this pull request. What's the next step in getting deferred resolving to work?

**Comment by acelot** [comment]
https://github.com/lordthorzonus/php-dataloader works for me very well. But my project based on ReactPHP. I have not tested data loader in regular PHP environment.

**Comment by Jalle19** [comment]
I still need to test this in a non-asynchronous environment before I dare merge this, I'm not super familiar with part of the code. I have to think if we really want to break the ExecutionInterface signature even though I said it was fine earlier.

**Comment by crisu83** [comment]
Hello @acelot. First of all, I'd like to thank you for all the work you've put into this. I'm going through a divorce so I haven't had any energy to focus on this project for quite some time now, but now I'm back at work after a long vacation so I'm picking it up slowly.

I agree that separate methods is the way to go, and we definitely want to hand over the promise from the public API, instead of resolving it before returning the result. This gives the consumer the freedom to do whatever they need with that promise.

I will now go through the changes and add my comments. After that, if all is good and no more changes are required we can merge this. Thanks once again for your efforts! 😊

**Comment by crisu83** [comment]
We had some problems with deferred resolvers. Any change you've had any success with them @acelot? The problem was that the promises got resolved too early, instead of using a dictionary, delaying the resolution and resolving them all at once.

File: src/api.php [hunk 1]
```diff
@@ -140,6 +142,50 @@ function execute(
...
+        $errorHandler
+    );
+
+    $data = null;
+    $resultPromise->then(function (ExecutionResult $result) use (&$data) {
+        $data = $result;
+    });
+    if ($data === null) {
+        $data = new ExecutionResult(null, [
+            new GraphQLException('It\'s looks like you are using Event Loop. Please use `executeAsync` method instead.')
```
Comment (crisu83) [inline src/api.php]: The message should be: 

```
Looks like you are using Event Loop. Please use `executeAsync` method instead.
```

File: src/api.php [hunk 2]
```diff
@@ -188,7 +234,58 @@ function graphql(
...
+        $errorHandler
+    );
+
+    $data = null;
+    $resultPromise->then(function (ExecutionResult $result) use (&$data) {
+        $data = $result;
+    });
+    if ($data === null) {
+        $data = new ExecutionResult(null, [
+            new GraphQLException('It\'s looks like you are using Event Loop. Please use `graphqlAsync` method instead.')
```
Comment (crisu83) [inline src/api.php]: The message should be: 

```
Looks like you are using Event Loop. Please use `graphqlAsync` method instead.
```

File: src/api.php [hunk 3]
```diff
@@ -188,7 +234,58 @@ function graphql(
...
+        $operationName,
+        $fieldResolver,
+        $errorHandler
+    );
+
+    $data = null;
+    $resultPromise->then(function (ExecutionResult $result) use (&$data) {
+        $data = $result;
+    });
+    if ($data === null) {
```
Comment (crisu83) [inline src/api.php]: I'd flip this condition to `null === $data` for consistency.

File: src/api.php [hunk 4]
```diff
@@ -140,6 +142,50 @@ function execute(
...
```
Comment (crisu83) [inline src/api.php]: I'd flip this condition to null === $data for consistency.

File: src/api.php [hunk 5]
```diff
@@ -140,6 +142,50 @@ function execute(
...
+        $document,
+        $rootValue,
+        $contextValue,
+        $variableValues,
+        $operationName,
+        $fieldResolver,
+        $errorHandler
+    );
+
+    $data = null;
```
Comment (crisu83) [inline src/api.php]: Linebreak after.

File: src/api.php [hunk 6]
```diff
@@ -140,6 +142,50 @@ function execute(
...
+    $resultPromise->then(function (ExecutionResult $result) use (&$data) {
+        $data = $result;
+    });
```
Comment (crisu83) [inline src/api.php]: Linebreak after.

File: src/api.php [hunk 7]
```diff
@@ -188,7 +234,58 @@ function graphql(
...
+        $source,
+        $rootValue,
+        $contextValue,
+        $variableValues,
+        $operationName,
+        $fieldResolver,
+        $errorHandler
+    );
+
+    $data = null;
```
Comment (crisu83) [inline src/api.php]: Linebreak after.

File: src/api.php [hunk 8]
```diff
@@ -188,7 +234,58 @@ function graphql(
...
+    $resultPromise->then(function (ExecutionResult $result) use (&$data) {
+        $data = $result;
+    });
```
Comment (crisu83) [inline src/api.php]: Linebreak after.

**Pull request merged**